pkg/pillar/cmd/client/client.go

```diff
@@ -284,25 +296,68 @@ func Run(ps *pubsub.PubSub) { //nolint:gocyclo
 				// XXX keep exponential unchanged?
 				break
 			}
-			// XXX server/tls setup from above to capture change
-			// to server file?
-			if operations["selfRegister"] {
-				done = selfRegister(zedcloudCtx, onboardTLSConfig, deviceCertPem, retryCount)
+
+			// if device is allowed to be V2, get server certs first unless the cloud does not support V2 or the
+			var delay time.Duration
+			certRetryCnt := 0
+			for !gotServerCerts && !cloudV1Only && devMaybeV2 {
+				time.Sleep(delay)
+				gotServerCerts, cloudV1Only = fetchCertChain(zedcloudCtx, devtlsConfig, certRetryCnt, true) // XXX always get certs from cloud for now
+				log.Infof("client fetchCertChain, gotServerCerts %v, v1only %v\n", gotServerCerts, cloudV1Only)
+				if cloudV1Only {
+					zedcloud.RemoveCertFile()
+					break
+				}
+				if gotServerCerts {
+					zedcloudCtx.V2API = true
+					continue
+				}
+				certRetryCnt++
+				if certRetryCnt > certsRetry { // XXX 3 retry for now
+					zedcloud.RemoveCertFile()
+					log.Errorf("Exceeded %d retries for fetchCertChain\n", certsRetry)
+					break
+				}
+				delay = 2 * (delay + time.Second)
+				if delay > maxDelay {
+					delay = maxDelay
+				}
+				log.Infof("Retrying fetchCertChain in %d seconds\n", delay/time.Second)
+			}
+
+			if !gotRegister && operations["selfRegister"] {
+				done, unknownAuth = selfRegister(zedcloudCtx, onboardTLSConfig, deviceCertPem, retryCount)
+				if done {
+					gotRegister = true
+				}
+				// if the https returns error of signed by unknown authority, we may have to yield
+				// to give a chance for Device Network Status to be updated for proxy certificates
+				if !done && unknownAuth && waitCount < waitDNSCnt {
+					waitCount++
+					log.Infof("client fetchCertChain, selfReg need wait proxy cert\n")
+					break
+				}
 				if !done && operations["getUuid"] {
 					// Check if getUUid succeeds
-					done, devUUID, hardwaremodel, enterprise, name = doGetUUID(zedcloudCtx, tlsConfig, retryCount)
+					done, _, devUUID, hardwaremodel, enterprise, name = doGetUUID(zedcloudCtx, devtlsConfig, retryCount)
 					if done {
 						log.Infof("getUUID succeeded; selfRegister no longer needed")
 						gotUUID = true
 					}
 				}
 			}
 			if !gotUUID && operations["getUuid"] {
-				done, devUUID, hardwaremodel, enterprise, name = doGetUUID(zedcloudCtx, tlsConfig, retryCount)
+				done, unknownAuth, devUUID, hardwaremodel, enterprise, name = doGetUUID(zedcloudCtx, devtlsConfig, retryCount)
 				if done {
 					log.Infof("getUUID succeeded; selfRegister no longer needed")
 					gotUUID = true
 				}
+				// yield for DNS update
+				if !done && unknownAuth && waitCount < waitDNSCnt {
+					waitCount++
+					log.Infof("client fetchCertChain, getUUID need wait proxy cert\n")
+					break
+				}
```

Comment:
You don't need this if using a timer instead of sleep. And that means that doGetUUID() doesn't need to return unknownauth either.

Reply:
ok